  Задача №2
  
Вы поднимаетесь по лестнице. Чтобы добраться до самого верха, вам необходимо пройти n ступеней. Каждый раз, вы можете подняться либо на 1 ступеньку, либо на 2. Как много различных способов подняться на самый верх у вас есть?

In [ ]:
def climb_stairs(n):
    if n <= 2:
        return n

    first, second = 1, 2

    for i in range(3, n + 1):
        first, second = second, first + second

    return second

# Тесты
print(climb_stairs(0))
print(climb_stairs(1))
print(climb_stairs(2))
print(climb_stairs(3))
print(climb_stairs(50))

0
1
2
3
20365011074


  Набор тестов, покрывающих все возможные краевые случаи:

Краевой случай с 0 ступенями: n = 0

Ожидаемый результат: 0


Краевой случай с 1 ступенью: n = 1

Ожидаемый результат: 1


Краевой случай с 2 ступенями: n = 2

Ожидаемый результат: 2


Обычный случай: n = 3

Ожидаемый результат: 3


Случай с большим количеством ступеней: n = 50

Ожидаемый результат: 20365011074

Для проверки взял число с сайта (ввёл 51, так как последовательность начинается с 0): https://geleot.ru/education/math/analytic_geometry/other/Fibonacci_number

  Алгоритмическая сложность

Временная сложность: O(n). Используем один цикл.

Пространственная сложность: O(1). Используем только две переменные для хранения чисел.

  Комментарий
  
Если бы понадобились повторные вычисления для различных значений n внутри одного приложения,
применение мемоизации или динамического программирования с кэшированием результатов могло бы улучшить производительность.